serverless/serverless
# Add support for Lambda@Edge

This is a rebased version of https://github.com/serverless/serverless/pull/4796 from @laardee 👍 🥇 🙌

## What did you implement:

Closes #3944 

Adds support for Lambda@Edge via pure CloudFormation. Thanks again to @laardee who jumped in to work on an initial implementation for this!

## How did you implement it:

Compile all the necessary CloudFormation resources (CloudFront / Versioned Lambda function, etc.) to add support for Lambda@Edge.

## Todos:

- [x] Retain Lambda Functions to avoid errors on Stack removals
- [x] Log info that functions have to be removed by hand when removing stack

## How can we verify it:

Create a new serverless service and replace the `handler.js` and `serverless.yml` code with the following snippets:

```javascript
// handler.js
module.exports.cloudfront = (event, context, callback) => {
  const response = event.Records[0].cf.response;
  const headers = response.headers;

  headers['x-serverless-time'] = [{ key: 'x-serverless-time', value: Date.now().toString() }];

  return callback(null, response);
};
```

```yaml
service: test-${self:custom.idx}

provider:
  name: aws
  runtime: nodejs10.x
  versionFunctions: false

custom:
  idx: 0

functions:
  hello:
    handler: handler.hello
    events:
      - cloudFront:
          eventType: viewer-response
          origin: https://serverless.com/framework/docs
```

Deploy the service (this could take a couple of minutes (~10 - 20mins)). After that you can access the CloudFront distribution URL which is printed in your console (you can also use `serverless info`). This should open up the Serverless Framework docs page.

Inspect the headers (e.g. using [this guide](https://www.mkyong.com/computer-tips/how-to-view-http-headers-in-google-chrome/) if you're running Chrome). You should see the `x-serverless-time` header which was added by the Lambda@Edge function.

Once validated you might want to remove the stack. Note the info message about the necessary manual Lambda@Edge function removals which is logged on the console.

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

       **Validate via `npm test`**
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO
